android/libraries/rib-router-navigator/src/main/kotlin/com/uber/rib/core/RouterAndState.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2023. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.rib.core
+
+import androidx.annotation.VisibleForTesting
+
+/**
+ * Internal class for keeping track of a navigation stack.
+ *
+ * @param state The associated state
+ * @param attachTransition The [RouterNavigator.AttachTransition] associated with this state.
+ * @param detachTransition The [RouterNavigator.DetachTransition] associated with this state.
+ * @param forceRouterCaching Override [RouterNavigatorState.isCacheable] behavior
+ */
+internal class RouterAndState<StateT : RouterNavigatorState>(
+  val state: StateT,
+  val attachTransition: RouterNavigator.AttachTransition<Router<*>, StateT>,
+  detachTransition: RouterNavigator.DetachTransition<Router<*>, StateT>?,
+  @get:VisibleForTesting val forceRouterCaching: Boolean = false
+) {
+  @get:Synchronized
+  @set:Synchronized
+  private var innerRouter: Router<*>? = null
+
+  /**
+   * Gets or creates the [Router] associated with this state. Router will be destroyed after
```

Comment:
will the router (thus the whole RIB) lose any state since this is no longer a strong reference?

Reply:
I think we must double-check not only going forward but **going back** on the stack for features that use this stacking with flag DEFAULT, to ensure no regressions are introduced.

Reply:
Correct, the `Router` instance will lose object state (changes in private properties). The right way to save any state is to use `Bundle` provided in `Router.dispatchAttach` (for state recovery) and `Router.saveInstanceState` (for the state saving). For the retro compatibility I've added new [isCacheable](https://github.com/uber/RIBs/pull/550/files#diff-5fb0dee6f8224afb283316fb92377bf3291abc13b217b0b8507f8733d7d08561) property, which is `false` by default (as recommended) + extended the `StackRouterNavigator` constructor by `forceRouterCaching` parameter which will override `isCacheable` by `true`, so after updating library we will have old behavior by default.